FairMQ device failed to change a state using FairMQ.SDK

**Describe the bug**
The DDS topology is started using DDS-control on a new Slurm cluster (`lxbk0595`). DDS-control hangs (or timeded out) trying to run `ConfigureRun` request (`.config` option of `sample-client`). Internally `dds-control-server` uses fair::mq::sdk::Topology::AsyncChangeState method to update the states of the devices in the topology.  Devices successfully updated to INIT DEVICE and COMPLETE INIT states but failed in the BIND state. 

From the device's logs we can see the following errors:
```
Error on handling DDS property update for data2: value=tcp://10.20.4.106:22497, senderTaskID=9051258523765084712: _Map_base::at
```

This is an exception in [fair::mq::plugins::DDS::SubscribeForConnectingChannels](https://github.com/FairRootGroup/FairMQ/blob/3f5374820a56c472a79e81b977861eddf4bff343/fairmq/plugins/DDS/DDS.cxx#L239)

For Sampler and Sink there is a different error in the logs:
```
[ERROR] Failed binding socket main/ProcessorGroup/ExCol_2/Sink_0.data2[0].pull, reason: Permission denied
```
But it looks like Sink and Sampler were able to recover.

**To Reproduce**
The test was done on a new Slurm cluster. You can login to it via interactive node `lxbk0595`.
DDS control is installed in `/lustre/rz/andrey/Soft/centos/DDS-control/install/`.
Run DDS control server:
```bash
bin/dds-control-server -r slurm -c /lustre/rz/andrey/test_slurm_job.sh 
```
Run client:
```bash
bin/sample-client --topo /lustre/rz/andrey/ex-dds-topology-infinite.xml 
```
The topology and Slurm config file you can see in the command options.

**Logs / Screenshots**
Full logs of the run can be found in `/lustre/rz/andrey/.DDS/sessions/950df43d-ca72-4655-a3dd-5ea1a209ddb5/2019-09-12-16-54-37`

`grep "\[ERROR\]" 2976_*/*.log` gives exact file names and line numbers with the error.

**System information (please complete the following information):**
 - OS: CentOS Linux 7 
 - Compiler: GCC 8.3.0
 - Environment: custom installation in `/lustre/rz/andrey/Soft/centos`

--- a/fairmq/plugins/DDS/DDS.cxx
+++ b/fairmq/plugins/DDS/DDS.cxx
@@ -8,8 +8,6 @@
 
 #include "DDS.h"
 
-#include <fairmq/sdk/commands/Commands.h>
-
 #include <fairmq/tools/Strings.h>
 
 #include <boost/algorithm/string/join.hpp>
@@ -315,116 +313,113 @@ auto DDS::SubscribeForCustomCommands() -> void
 
     fDDS.SubscribeCustomCmd([id, this](const string& cmdStr, const string& cond, uint64_t senderId) {
         // LOG(info) << "Received command: '" << cmdStr << "' from " << senderId;
-
-        using namespace fair::mq::sdk;
-        cmd::Cmds inCmds;
+        sdk::cmd::Cmds inCmds;
         inCmds.Deserialize(cmdStr);
-
         for (const auto& cmd : inCmds) {
-            // LOG(info) << "Received command type: '" << cmd->GetType() << "' from " << senderId;
-            switch (cmd->GetType()) {
-                case cmd::Type::check_state: {
-                    fDDS.Send(cmd::Cmds(cmd::make<cmd::CurrentState>(id, GetCurrentDeviceState()))
-                                  .Serialize(),
-                              to_string(senderId));
-                } break;
-                case cmd::Type::change_state: {
-                    Transition transition = static_cast<cmd::ChangeState&>(*cmd).GetTransition();
-                    if (ChangeDeviceState(transition)) {
-                        cmd::Cmds outCmds(
-                            cmd::make<cmd::TransitionStatus>(id, fDDSTaskId, cmd::Result::Ok, transition));
-                        fDDS.Send(outCmds.Serialize(), to_string(senderId));
-                    } else {
-                        sdk::cmd::Cmds outCmds(
-                            cmd::make<cmd::TransitionStatus>(id, fDDSTaskId, cmd::Result::Failure, transition));
-                        fDDS.Send(outCmds.Serialize(), to_string(senderId));
-                    }
-                    {
-                        lock_guard<mutex> lock{fStateChangeSubscriberMutex};
-                        fLastExternalController = senderId;
-                    }
-                } break;
-                case cmd::Type::dump_config: {
-                    stringstream ss;
-                    for (const auto pKey : GetPropertyKeys()) {
-                        ss << id << ": " << pKey << " -> " << GetPropertyAsString(pKey) << "\n";
-                    }
-                    cmd::Cmds outCmds(cmd::make<cmd::Config>(id, ss.str()));
-                    fDDS.Send(outCmds.Serialize(), to_string(senderId));
-                } break;
-                case cmd::Type::state_change_exiting_received: {
-                    {
-                        lock_guard<mutex> lock{fStateChangeSubscriberMutex};
-                        if (fLastExternalController == senderId) {
-                            fExitingAckedByLastExternalController = true;
-                        }
-                    }
-                    fExitingAcked.notify_one();
-                } break;
-                case cmd::Type::subscribe_to_state_change: {
-                    lock_guard<mutex> lock{fStateChangeSubscriberMutex};
-                    fStateChangeSubscribers.insert(senderId);
-
-                    LOG(debug) << "Publishing state-change: " << fLastState << "->" << fCurrentState
-                               << " to " << senderId;
-
-                    cmd::Cmds outCmds(
-                        cmd::make<cmd::StateChangeSubscription>(id, cmd::Result::Ok),
-                        cmd::make<cmd::StateChange>(id, fDDSTaskId, fLastState, fCurrentState));
-
-                    fDDS.Send(outCmds.Serialize(), to_string(senderId));
-                } break;
-                case cmd::Type::unsubscribe_from_state_change: {
-                    {
-                        lock_guard<mutex> lock{fStateChangeSubscriberMutex};
-                        fStateChangeSubscribers.erase(senderId);
-                    }
-                    cmd::Cmds outCmds(
-                        cmd::make<cmd::StateChangeUnsubscription>(id, cmd::Result::Ok));
-                    fDDS.Send(outCmds.Serialize(), to_string(senderId));
-                } break;
-                case cmd::Type::get_properties: {
-                    auto _cmd = static_cast<cmd::GetProperties&>(*cmd);
-                    auto const request_id(_cmd.GetRequestId());
-                    auto result(cmd::Result::Ok);
-                    std::vector<std::pair<std::string, std::string>> props;
-                    try {
-                        for (auto const& prop : GetPropertiesAsString(_cmd.GetQuery())) {
-                            props.push_back({prop.first, prop.second});
-                        }
-                    } catch (std::exception const& e) {
-                        LOG(warn) << "Getting properties (request id: " << request_id << ") failed: " << e.what();
-                        result = cmd::Result::Failure;
-                    }
-                    cmd::Cmds const outCmds(cmd::make<cmd::Properties>(id, request_id, result, props));
-                    fDDS.Send(outCmds.Serialize(), to_string(senderId));
-                } break;
-                case cmd::Type::set_properties: {
-                    auto _cmd(static_cast<cmd::SetProperties&>(*cmd));
-                    auto const request_id(_cmd.GetRequestId());
-                    auto result(cmd::Result::Ok);
-                    try {
-                        fair::mq::Properties props;
-                        for (auto const& prop : _cmd.GetProps()) {
-                            props.insert({prop.first, fair::mq::Property(prop.second)});
-                        }
-                        // TODO Handle builtin keys with different value type than string
-                        SetProperties(props);
-                    } catch (std::exception const& e) {
-                        LOG(warn) << "Setting properties (request id: " << request_id << ") failed: " << e.what();
-                        result = cmd::Result::Failure;
-                    }
-                    cmd::Cmds const outCmds(cmd::make<cmd::PropertiesSet>(id, request_id, result));
-                    fDDS.Send(outCmds.Serialize(), to_string(senderId));
-                } break;
-                default:
-                    LOG(warn) << "Unexpected/unknown command received: " << cmdStr;
-                    LOG(warn) << "Origin: " << senderId;
-                    LOG(warn) << "Destination: " << cond;
-                break;
-            }
+            HandleCmd(id, *cmd, cond, senderId);
         }
     });
+}
+
+auto DDS::HandleCmd(const string& id, sdk::cmd::Cmd& cmd, const string& cond, uint64_t senderId) -> void
+{
+    using namespace fair::mq::sdk;
+    using namespace fair::mq::sdk::cmd;
+    // LOG(info) << "Received command type: '" << cmd.GetType() << "' from " << senderId;
+    switch (cmd.GetType()) {
+        case Type::check_state: {
+            fDDS.Send(Cmds(make<CurrentState>(id, GetCurrentDeviceState())).Serialize(), to_string(senderId));
+        } break;
+        case Type::change_state: {
+            Transition transition = static_cast<ChangeState&>(cmd).GetTransition();
+            if (ChangeDeviceState(transition)) {
+                Cmds outCmds(make<TransitionStatus>(id, fDDSTaskId, Result::Ok, transition));
+                fDDS.Send(outCmds.Serialize(), to_string(senderId));
+            } else {
+                Cmds outCmds(make<TransitionStatus>(id, fDDSTaskId, Result::Failure, transition));
+                fDDS.Send(outCmds.Serialize(), to_string(senderId));
+            }
+            {
+                lock_guard<mutex> lock{fStateChangeSubscriberMutex};
+                fLastExternalController = senderId;
+            }
+        } break;
+        case Type::dump_config: {
+            stringstream ss;
+            for (const auto pKey : GetPropertyKeys()) {
+                ss << id << ": " << pKey << " -> " << GetPropertyAsString(pKey) << "\n";
+            }
+            Cmds outCmds(make<Config>(id, ss.str()));
+            fDDS.Send(outCmds.Serialize(), to_string(senderId));
+        } break;
+        case Type::state_change_exiting_received: {
+            {
+                lock_guard<mutex> lock{fStateChangeSubscriberMutex};
+                if (fLastExternalController == senderId) {
+                    fExitingAckedByLastExternalController = true;
+                }
+            }
+            fExitingAcked.notify_one();
+        } break;
+        case Type::subscribe_to_state_change: {
+            lock_guard<mutex> lock{fStateChangeSubscriberMutex};
+            fStateChangeSubscribers.insert(senderId);
+
+            LOG(debug) << "Publishing state-change: " << fLastState << "->" << fCurrentState << " to " << senderId;
+
+            Cmds outCmds(make<StateChangeSubscription>(id, Result::Ok),
+                         make<StateChange>(id, fDDSTaskId, fLastState, fCurrentState));
+
+            fDDS.Send(outCmds.Serialize(), to_string(senderId));
+        } break;
+        case Type::unsubscribe_from_state_change: {
+            {
+                lock_guard<mutex> lock{fStateChangeSubscriberMutex};
+                fStateChangeSubscribers.erase(senderId);
+            }
+            Cmds outCmds(make<StateChangeUnsubscription>(id, Result::Ok));
+            fDDS.Send(outCmds.Serialize(), to_string(senderId));
+        } break;
+        case Type::get_properties: {
+            auto _cmd = static_cast<cmd::GetProperties&>(cmd);
+            auto const request_id(_cmd.GetRequestId());
+            auto result(Result::Ok);
+            std::vector<std::pair<std::string, std::string>> props;
+            try {
+                for (auto const& prop : GetPropertiesAsString(_cmd.GetQuery())) {
+                    props.push_back({prop.first, prop.second});
+                }
+            } catch (std::exception const& e) {
+                LOG(warn) << "Getting properties (request id: " << request_id << ") failed: " << e.what();
+                result = Result::Failure;
+            }
+            Cmds const outCmds(make<cmd::Properties>(id, request_id, result, props));
+            fDDS.Send(outCmds.Serialize(), to_string(senderId));
+        } break;
+        case Type::set_properties: {
+            auto _cmd(static_cast<cmd::SetProperties&>(cmd));
+            auto const request_id(_cmd.GetRequestId());
+            auto result(Result::Ok);
+            try {
+                fair::mq::Properties props;
+                for (auto const& prop : _cmd.GetProps()) {
+                    props.insert({prop.first, fair::mq::Property(prop.second)});
+                }
+                // TODO Handle builtin keys with different value type than string
+                SetProperties(props);
+            } catch (std::exception const& e) {
+                LOG(warn) << "Setting properties (request id: " << request_id << ") failed: " << e.what();
+                result = Result::Failure;
+            }
+            Cmds const outCmds(make<PropertiesSet>(id, request_id, result));
+            fDDS.Send(outCmds.Serialize(), to_string(senderId));
+        } break;
+        default:
+            LOG(warn) << "Unexpected/unknown command received: " << cmd.GetType();
+            LOG(warn) << "Origin: " << senderId;
+            LOG(warn) << "Destination: " << cond;
+        break;
+            }
 }
 
 DDS::~DDS()

--- a/fairmq/plugins/DDS/DDS.cxx
+++ b/fairmq/plugins/DDS/DDS.cxx
@@ -283,7 +283,10 @@ auto DDS::SubscribeForConnectingChannels() -> void
                     if (mi.second.fNumSubChannels == mi.second.fDDSValues.size()) {
                         int i = 0;
                         for (const auto& e : mi.second.fDDSValues) {
-                            SetProperty<string>(string{"chans." + mi.first + "." + to_string(i) + ".address"}, e.second);
+                            auto result = UpdateProperty<string>(string{"chans." + mi.first + "." + to_string(i) + ".address"}, e.second);
+                            if (!result) {
+                                LOG(error) << "UpdateProperty failed for: " << "chans." << mi.first << "." << to_string(i) << ".address" << " - property does not exist";
+                            }
                             ++i;
                         }
                     }

--- a/fairmq/plugins/DDS/DDS.cxx
+++ b/fairmq/plugins/DDS/DDS.cxx
@@ -243,6 +243,12 @@ auto DDS::SubscribeForConnectingChannels() -> void
                     unique_lock<mutex> lk(fUpdateMutex);
                     fUpdateCondition.wait(lk, [&]{ return fUpdatesAllowed; });
                 }
+
+                if (fConnectingChans.find(channelName) == fConnectingChans.end()) {
+                    LOG(error) << "Received an update for a connecting channel, but either no channel with given channel name exists or it has already been configured: '" << channelName << "', ignoring...";
+                    return;
+                }
+
                 string val = value;
                 // check if it is to handle as one out of multiple values
                 auto it = fIofN.find(channelName);

--- a/fairmq/plugins/DDS/DDS.cxx
+++ b/fairmq/plugins/DDS/DDS.cxx
@@ -37,6 +37,7 @@ DDS::DDS(const string& name,
          const string& homepage,
          PluginServices* pluginServices)
     : Plugin(name, version, maintainer, homepage, pluginServices)
+    , fDDSTaskId(dds::env_prop<dds::task_id>())
     , fCurrentState(DeviceState::Idle)
     , fLastState(DeviceState::Idle)
     , fDeviceTerminationRequested(false)
@@ -112,7 +113,7 @@ DDS::DDS(const string& name,
             for (auto subscriberId : fStateChangeSubscribers) {
                 LOG(debug) << "Publishing state-change: " << fLastState << "->" << newState << " to " << subscriberId;
 
-                Cmds cmds(make<StateChange>(id, dds::env_prop<dds::task_id>(), fLastState, fCurrentState));
+                Cmds cmds(make<StateChange>(id, fDDSTaskId, fLastState, fCurrentState));
                 fDDS.Send(cmds.Serialize(), to_string(subscriberId));
             }
         });
@@ -143,10 +144,10 @@ auto DDS::StartWorkerThread() -> void
 auto DDS::WaitForExitingAck() -> void
 {
     unique_lock<mutex> lock(fStateChangeSubscriberMutex);
-    fExitingAcked.wait_for(
-        lock,
-        chrono::milliseconds(GetProperty<unsigned int>("wait-for-exiting-ack-timeout")),
-        [this]() { return fExitingAckedByLastExternalController; });
+    auto timeout = GetProperty<unsigned int>("wait-for-exiting-ack-timeout");
+    fExitingAcked.wait_for(lock, chrono::milliseconds(timeout), [this]() {
+        return fExitingAckedByLastExternalController;
+    });
 }
 
 auto DDS::FillChannelContainers() -> void
@@ -321,11 +322,11 @@ auto DDS::SubscribeForCustomCommands() -> void
                     Transition transition = static_cast<cmd::ChangeState&>(*cmd).GetTransition();
                     if (ChangeDeviceState(transition)) {
                         cmd::Cmds outCmds(
-                            cmd::make<cmd::TransitionStatus>(id, dds::env_prop<dds::task_id>(), cmd::Result::Ok, transition));
+                            cmd::make<cmd::TransitionStatus>(id, fDDSTaskId, cmd::Result::Ok, transition));
                         fDDS.Send(outCmds.Serialize(), to_string(senderId));
                     } else {
                         sdk::cmd::Cmds outCmds(
-                            cmd::make<cmd::TransitionStatus>(id, dds::env_prop<dds::task_id>(), cmd::Result::Failure, transition));
+                            cmd::make<cmd::TransitionStatus>(id, fDDSTaskId, cmd::Result::Failure, transition));
                         fDDS.Send(outCmds.Serialize(), to_string(senderId));
                     }
                     {
@@ -359,8 +360,7 @@ auto DDS::SubscribeForCustomCommands() -> void
 
                     cmd::Cmds outCmds(
                         cmd::make<cmd::StateChangeSubscription>(id, cmd::Result::Ok),
-                        cmd::make<cmd::StateChange>(
-                            id, dds::env_prop<dds::task_id>(), fLastState, fCurrentState));
+                        cmd::make<cmd::StateChange>(id, fDDSTaskId, fLastState, fCurrentState));
 
                     fDDS.Send(outCmds.Serialize(), to_string(senderId));
                 } break;

--- a/fairmq/plugins/DDS/DDS.cxx
+++ b/fairmq/plugins/DDS/DDS.cxx
@@ -74,15 +74,22 @@ DDS::DDS(const string& name,
         // subscribe to device state changes, pushing new state changes into the event queue
         SubscribeToDeviceStateChange([&](DeviceState newState) {
             switch (newState) {
-                case DeviceState::Bound:
+                case DeviceState::Bound: {
                     // Receive addresses of connecting channels from DDS
                     // and propagate addresses of bound channels to DDS.
                     FillChannelContainers();
 
+                    // allow updates from key value after channel containers are filled
+                    {
+                        lock_guard<mutex> lk(fUpdateMutex);
+                        fUpdatesAllowed = true;
+                    }
+                    fUpdateCondition.notify_one();
+
                     // publish bound addresses via DDS at keys corresponding to the channel
                     // prefixes, e.g. 'data' in data[i]
                     PublishBoundChannels();
-                    break;
+                } break;
                 case DeviceState::ResettingDevice: {
                     {
                         lock_guard<mutex> lk(fUpdateMutex);
@@ -90,9 +97,8 @@ DDS::DDS(const string& name,
                     }
 
                     EmptyChannelContainers();
-                    break;
-                }
-                case DeviceState::Exiting:
+                } break;
+                case DeviceState::Exiting: {
                     if (!fControllerThread.joinable()) {
                         fControllerThread = thread(&DDS::WaitForExitingAck, this);
                     }
@@ -100,7 +106,7 @@ DDS::DDS(const string& name,
                     fDeviceTerminationRequested = true;
                     UnsubscribeFromDeviceStateChange();
                     ReleaseDeviceControl();
-                    break;
+                } break;
                 default:
                     break;
             }
@@ -213,11 +219,6 @@ auto DDS::FillChannelContainers() -> void
             LOG(debug) << "dds-i-n: adding " << chanName << " -> i: " << i << " n: " << n;
             fIofN.insert(make_pair(chanName, IofN(i, n)));
         }
-        {
-            lock_guard<mutex> lk(fUpdateMutex);
-            fUpdatesAllowed = true;
-        }
-        fUpdateCondition.notify_one();
     } catch (const exception& e) {
         LOG(error) << "Error filling channel containers: " << e.what();
     }

--- a/fairmq/plugins/DDS/DDS.h
+++ b/fairmq/plugins/DDS/DDS.h
@@ -144,6 +144,7 @@ class DDS : public Plugin
     auto SubscribeForCustomCommands() -> void;
 
     DDSSubscription fDDS;
+    size_t fDDSTaskId;
 
     std::unordered_map<std::string, std::vector<std::string>> fBindingChans;
     std::unordered_map<std::string, DDSConfig> fConnectingChans;

--- a/fairmq/plugins/DDS/DDS.h
+++ b/fairmq/plugins/DDS/DDS.h
@@ -12,6 +12,7 @@
 #include <fairmq/Plugin.h>
 #include <fairmq/StateQueue.h>
 #include <fairmq/Version.h>
+#include <fairmq/sdk/commands/Commands.h>
 
 #include <dds/dds.h>
 
@@ -142,6 +143,7 @@ class DDS : public Plugin
     auto SubscribeForConnectingChannels() -> void;
     auto PublishBoundChannels() -> void;
     auto SubscribeForCustomCommands() -> void;
+    auto HandleCmd(const std::string& id, sdk::cmd::Cmd& cmd, const std::string& cond, uint64_t senderId) -> void;
 
     DDSSubscription fDDS;
     size_t fDDSTaskId;

